src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor.go

```diff
@@ -392,6 +389,8 @@ func (spa *StoragePowerActor_I) Surprise(ticket block.Ticket) []addr.Address {
 		// SendMessage(sm.ExtendProvingPeriod)
 	}
 
+	h.Commit(st)
```

Comment:
- i think this is only reading (not committing), so this should be `h.Release(st)` ?
- given the state handle, the `Send` function will fail inside, will want to move it to a loop like this:

```go
h, st := a.State(rt)
// make all the variables

index := 0
for address, _ := range st.PowerTable() {
    minerAddresses[index] = address
    index++
}

for i := uint(0); i < challengeCount; i++ {
  var randomNumber uint
  minerIndex := randomNumber % uint(len(st.PowerTable()))
  minerAddress := minerAddresses[minerIndex]
  surprisedMiners = append(surprisedMiners, minerAddress)
}

h.Release()


// now send the messages (after state is released)
for _, addr := range surprisedMiners {
  rt.SendMessage(addr, ...)
}
```

Reply:
oh and potentially also:

```go
func (a *StoragePowerActorCode_I) Surprise(rt Runtime, ticket block.Ticket) []addr.Address {

  var provingPeriod int // TODO

  // sample the actor addresses
  h, st := a.State(rt)
  challengeCount := len(st.PowerTable()) / provingPeriod
  surprisedMiners := st.SampleMinersToSurprise(rt, challengeCount)
  st.Release() 

  // now send the messages
  for _, addr := range surprisedMiners {
    rt.SendMessage(addr, ...)
  }
}


func (a *StoragePowerActorState_I) SampleMinersToSurprise(rt Runtime, challengeCount int) []addr.Address {
  // this wont quite work -- a.PowerTable() is a HAMT by actor address, doesn't
  // support enumerating by int index. maybe we need that as an interface too, 
  // or something similar to an iterator (or iterator over the keys)
  // or even a seeded random call directly in the HAMT: myhamt.GetRandomElement(seed []byte, idx int)
  m := a.PowerTable() 

  return postSurpriseSample(rt, m, challengeCount)
}

// postSurpriseIndices implements the PoSt-Surprise sampling algorithm
func postSurpriseSample(rt Runtime, allMiners []addr.Address, provingPeriod int) []addr.Address {

  sm := make([]addr.Address, challengeCount)
  for i := 0; i < challengeCount; i++ {
    // rInt := rt.NextRandomInt() // we need something like this in the runtime
    rInt := 4 // xkcd prng for now.
    miner := allMiners[rInt % len(allMiners)]
    sm = append(sm, miner)
  }
  return sm
}
```

Reply:
This is way cleaner than before 👍, will add this as placeholder and revisit in a later PR.